packages/apollo-language-server/src/__tests__/diagnostics.ts

```diff
@@ -0,0 +1,81 @@
+import { Source } from "graphql";
+import { loadSchema } from "apollo-codegen-core/lib/loading";
+import { GraphQLDocument } from "../document";
+import { collectExecutableDefinitionDiagnositics } from "../diagnostics";
+
+const schema = loadSchema(
+  require.resolve("../../../common-test/fixtures/starwars/schema.json")
+);
+
+const validDocument = new GraphQLDocument(
+  new Source(`
+    query HeroAndFriendsNames {
+      hero {
+        name
+        friends {
+          name
+        }
+      }
+    }`)
+);
+
+const invalidDocument = new GraphQLDocument(
+  new Source(`
+    query HeroAndFriendsNames {
+      hero {
+        nam         # Missing letter 'e'
+        friend {    # Missing letter 's'
+          name
+        }
+      }
+    }`)
+);
+
+const documentWithTypes = new GraphQLDocument(
+  new Source(`
+    type SomeType {
+      thing: String
+    }
+
+    enum SomeEnum {
+      THING_ONE
+      THING_TWO
+    }
+
+    query HeroAndFriendsNames {
+      hero {
+        name
+        friends {
+          name
+        }
+      }
+    }`)
+);
+
+describe("Language server diagnostics", () => {
+  describe("collectExecutableDefinitionDiagnositics", () => {
+    test("Returns no diagnostics for a correct document", () => {
```

Comment:
Thanks for taking the time to write tests, we should definitely do more of that!

In terms of style, I personally prefer what is sometimes called spec style, where you use `it` instead of `test` and the combined message reads as a description of behavior. That actually already works well here: `it("returns no diagnostics for a correct document"` would read as 'collectExecutableDefinitionDiagnositics returns no diagnostics for a correct document'.

Reply:
We don't have real testing guidelines in place unfortunately, so that might be something we'd want to do. If you're looking for examples of tests, you could look [at this file](https://github.com/apollographql/apollo-cli/blob/master/packages/apollo-codegen-core/src/compiler/visitors/__tests__/conditionalFields.ts) and other tests that are part of codegen. One nice addition there is the use of a custom matcher `toMatchSelectionSet` (see [here](https://github.com/apollographql/apollo-cli/blob/master/packages/apollo-codegen-core/test-utils/matchers.ts)). Custom matchers can make both code and failure output a lot more readable (and therefore actionable). There's also a growing collection of [built-in matchers](https://jestjs.io/docs/en/expect), so you may want to have a look there.

Reply:
Thanks for showing me these custom matchers - are you suggesting they have an application here? I looked into them a bit and it didn't seem quite what I was looking for, since I'm just confirming that validation messages come back. I could do a direct string comparison on the messages if you prefer the test be more specific than "confirm there are 2 errors in the result"?